### PR TITLE
Incomplete Beta Function Inverse

### PR DESCRIPTION
**Pull request by andrjohns**
## Summary
This PR adds the inverse of the incomplete beta function, will be needed for some of the quantile/icdf functions to come.

## Tests
Individual `rev/fwd/mix` tests have been added, as the gradients use the `F32` hypergeometric series and the autodiff finite-differencing tests are not stable.

## Side Effects

The `F32` templating has been relaxed to allow each of the inputs to be different types

## Release notes

Add the incomplete beta function inverse

## Checklist

- [x] Math issue #2513 

- [x] Copyright holder: Andrew Johnson

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/rev/fun/inc_beta_inv.hpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#include <stan/math/prim/fun/is_any_nan.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * The inverse of the normalized incomplete beta function of a, b, with
+ * probability p.
+ *
+ * Used to compute the cumulative density function for the beta
```
Comment (spinkney) [inline stan/math/rev/fun/inc_beta_inv.hpp:25]: Used to compute the _inverse_ cumulative density function

**Comment by spinkney** [comment]
@andrjohns do you mind if I put in the latex formulas for the partials in the rev file documentation?

**Comment by andrjohns** [comment]
Ah good point forgot about those! Yeah feel free to add them, otherwise I can do it tonight/tomorrow

**Comment by andrjohns** [comment]
@serban-nicusor-toptal it looks like the downstream tests for this are failing due to some odd git errors, would you mind having a look when you get a minute?

**Comment by serban-nicusor-toptal** [comment]
Hey, checking!

**Comment by andrjohns** [comment]
Thanks for sorting this so quickly!

**Comment by spinkney** [comment]
I don't feel comfortable reviewing the cpp but the derivatives and tests look good to me. Maybe @SteveBronder can take a look?

**Comment by spinkney** [comment]
will it be possible to get this in to the 2.29 release? We can then do `student_t` copulas

**Comment by andrjohns** [comment]
@serban-nicusor-toptal looks like the jenkins for this one is having trouble with `clang-format`, if you wouldn't mind having a look when you've got a minute:
```
git checkout -b feature/ibeta_inv
Switched to a new branch 'feature/ibeta_inv'
+ clang-format --version
/home/jenkins/workspace/Stan_Math_PR-2637@tmp/durable-c616a978/script.sh: line 4: clang-format: command not found
+ xargs -n20 -P8 clang-format -i
+ find stan test -name '*.hpp' -o -name '*.cpp'
xargs: clang-format: No such file or directory
++ git diff
+ [[ '' != '' ]]
```

**Comment by rok-cesnovar** [comment]
Can you try merging in the latest develop, there were quite a few changes in the Jenkinsfile and test machine since then. Hopefully that will help.

**Comment by andrjohns** [comment]
> Can you try merging in the latest develop, there were quite a few changes in the Jenkinsfile and test machine since then. Hopefully that will help.

Ah thought I had but it hadn't triggered a new CI run for some reason

**Comment by andrjohns** [comment]
@rok-cesnovar or @SteveBronder would either of you have a minute to review this one? @spinkney has signed off on the math, so it's just the c++ to be checked

**Comment by rok-cesnovar** [comment]
I can do it today. We wont be able to merge after we do the patch release (https://github.com/stan-dev/cmdstan/issues/1086) which should be in a day or two.

**Review by rok-cesnovar — COMMENTED** [review]
Looks great. I only have one question in two places.

File: stan/math/prim/fun/inv_inc_beta.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * The inverse of the normalized incomplete beta function of a, b, with
+ * probability p.
+ *
+ * Used to compute the inverse cumulative density function for the beta
+ * distribution.
+ *
+ * @param a Shape parameter a >= 0; a and b can't both be 0
+ * @param b Shape parameter b >= 0
+ * @param p Random variate. 0 <= p <= 1
```
[the remark this answers is not in the document]
Comment (andrjohns) [reply stan/math/prim/fun/inv_inc_beta.hpp:21]: Right you are! Added a check to the `prim` implementation, which gets called by the autodiff types as well

**Comment by andrjohns** [comment]
@rok-cesnovar This should be all ready for another look

**Review by rok-cesnovar — APPROVED** [review]
Good to go! Thanks Andrew!

**Comment by andrjohns** [comment]
Thanks!

**Pull request merged**